### ISSUE TITLE
CAS upload speeds optimizations

### ISSUE THREAD

**Issue by migmartri** [chainloop-dev/chainloop#375]
The CAS proxy works as an intermediary between the Chainloop Client (Operator CLI, Attestation CLI , Web UI) and the final destination backend (OCI, S3, Azure Blob Storage). 

![CAS Sharing](https://github.com/chainloop-dev/chainloop/assets/24523/fadfde59-72d6-4266-ac2b-be30edbbe1e4)

We've noticed that in the default setup [1] upload speeds are consistently +10x slower than downloads.

```
$ chainloop artifact upload -f Downloads/spring-petclinic-3.1.0-SNAPSHOT.jar
INF uploading spring-petclinic-3.1.0-SNAPSHOT.jar - sha256:3758abd13c9edc08fc4ac345e07fb6fdf0e3cf8d596534e8d4b89a045e5c049b
 ... done! [58.18MB in 1m50.571s; 526.19KB/s]
```

```
chainloop artifact download -d sha256:3758abd13c9edc08fc4ac345e07fb6fdf0e3cf8d596534e8d4b89a045e5c049b
INF downloading file name=spring-petclinic-3.1.0-SNAPSHOT.jar to=/home/migmartri/spring-petclinic-3.1.0-SNAPSHOT.jar
99.97% [#################.] [58.16MB in 6.383549s; 9.11MB/s] ...
```

We should analyze and tweak the setup to improve such upload times. Nginx Ingress configuration, ELB configuration upload code itself, and of course k8s resource allocation.

[1] 

- Chainloop deployed in Kubernetes
- Nginx Ingress Controller behind an ELB

**Comment by migmartri** [comment]
I've done some testing and these are the conclusions

### It's not the NGINX ingress controller.

I've deployed Chainloop without nginx ingress controller and using AWS ALB directly and the speed is more or less the same

Also running an upload from AWS the upload is fast 50MB

```
chainloop config view
Config file: /root/.config/chainloop/config.toml
control-plane: api.cp.chainloop.dev:443
artifact-cas: api.cas.chainloop.dev:443

chainloop artifact upload -f debian-12.1.0-amd64-netinst.iso 
INF uploading debian-12.1.0-amd64-netinst.iso - sha256:9f181ae12b25840a508786b1756c6352a0e58484998669288c4eec2ab16b8559
13.88% [##................] [91.23MB in 1.949526s; ~ETA: 13s; 46.79MB/s] .
```

Additionally, I've migrated to network load balancer instead of the classic one and it does affect performance.

**Comment by migmartri** [comment]
To make sure this is not a result of internal networking routing I did the test also in an Azure Kubernetes Cluster with good results too

```
14.83% [##................] [97.52MB in 3.410847s; ~ETA: 21s; 28.59MB/s] ... 
```

The key here is that this cluster is also in the same region US-EAST

**Comment by migmartri** [comment]
It looks a lot like this issue https://github.com/grpc/grpc-go/issues/1043

Although in theory this was already fixed https://grpc.io/blog/grpc-go-perf-improvements/

**Comment by migmartri** [comment]
I ran some tests locally simulating latency with [latency.go](https://github.com/grpc/grpc-go/blob/master/benchmark/latency/latency.go), and both the dynamic control flow and the static one seemed to work as expected.

But once I test it behind both an application load balancer (AWS ALB) and a network load balancer (AWS NLB) + nginx the problem persists.

**I tried to upload directly doing port-forward and the speed seems correct, so there might be a miss-configuration in nginx / ALB**

```
INF uploading tst - sha256:49bc20df15e412a64472421e13fe86ff1c5165e18b2afccf160d4dc19fe68a14
10.68% [#.................] [114.71MB in 11.177917s; ~ETA: 1m41s; 10.26MB/s] ... 
```

This blog post might shed some light on this https://blog.cloudflare.com/delivering-http-2-upload-speed-improvements/

**Comment by migmartri** [comment]
`client-body-buffer-size` in nginx fixes the issue :dropmic

**Comment by migmartri** [comment]
To summarize my findings and how it was fixed.

- Artifacts uploads (both during attestation and manual uploads) were slow when there was over 50ms in RTT latency. For that operation, we use gRPC streaming over http2.
- To pinpoint the issue, I configured a deployment using instead of NGINX ingress controller + l4 load balancer an AWS Application Load balancer (L7). The uploads were still slow.
- I looked into go-grpc code to see if there was any sign of trouble, and [there was](https://github.com/grpc/grpc-go/issues/1043). It points to an issue with how http2 does flow control/buffering. But the issue [was fixed long time ago](https://grpc.io/blog/grpc-go-perf-improvements/). I, in fact, did some tests with the neat [latency.go](https://github.com/grpc/grpc-go/blob/master/benchmark/latency/latency.go) library and I could attest that dynamic windows were working as expected.
- I ran some testing with port-forwarding directly to the CAS pod, and surprise! The uploads were fast, so the issue was indeed in Nginx frontend.
- And the solution was pointed out by [this blog post](https://blog.cloudflare.com/delivering-http-2-upload-speed-improvements/), in which you can see that, in fact, Nginx as reverse proxy runs its own http2 flow-control. Increasing the buffer fixed the issue #381.
- But why didn't it work either with AWS ALB? Well, it seems that AWS ALB has the same lower bound buffer value, but I can't find a way to configure it.


For the future

- flow control in grpc-go also allows configuring the initial window which could improve the initial upload burst https://pkg.go.dev/google.golang.org/grpc@v1.57.0#InitialWindowSize
- This issue might affect also GCP load balancer, we need to take a look at it. For now, we can only guarantee that uploads are efficient with Nginx Ingress.

cc/ @jiparis, @jotadrilo, @zaibon and @danlishka
